File: basic_dsa_ninety_five.java
/**
 * Basic DSA Problems - Majority Element
 * Simple and essential DSA problems for beginners
 */
public class basic_dsa_ninety_five {

    // Problem: Find the majority element (appears more than n/2 times). Assume it always exists.
    // Example: [3,2,3] -> 3, [2,2,1,1,1,2,2] -> 2

    /**
     * Boyer-Moore voting: one candidate, count; majority cancels all others.
     * Time O(n), Space O(1)
     */
    public static int majorityElement(int[] nums) {
        if (nums == null || nums.length == 0) return -1;
        int cand = nums[0], count = 1;
        for (int i = 1; i < nums.length; i++) {
            if (count == 0) cand = nums[i];
            count += (nums[i] == cand) ? 1 : -1;
        }
        return cand;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Majority Element");
        System.out.println("=".repeat(60));

        System.out.println("[3,2,3] -> " + majorityElement(new int[] { 3, 2, 3 }));
        System.out.println("[2,2,1,1,1,2,2] -> " + majorityElement(new int[] { 2, 2, 1, 1, 1, 2, 2 }));
    }
}

File: basic_dsa_ninety_five.cpp
/**
 * Basic DSA - Number of Good Pairs (LeetCode 1512)
 * Count pairs (i,j) where i<j and nums[i]==nums[j].
 */
#include <iostream>
#include <vector>
#include <unordered_map>

int numIdenticalPairs(std::vector<int>& nums) {
    std::unordered_map<int, int> count;
    int pairs = 0;
    for (int x : nums) {
        pairs += count[x];
        count[x]++;
    }
    return pairs;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Number of Good Pairs\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {1, 2, 3, 1, 1, 3};
    std::cout << "[1,2,3,1,1,3] -> " << numIdenticalPairs(a1) << "\n";

    std::vector<int> a2 = {1, 1, 1, 1};
    std::cout << "[1,1,1,1] -> " << numIdenticalPairs(a2) << "\n";

    std::vector<int> a3 = {1, 2, 3};
    std::cout << "[1,2,3] -> " << numIdenticalPairs(a3) << "\n";
}
